public/js/History.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * History: Keep track of user's sample history.
+ *
+ * This object emulates the typical back/forward behaviour in browsers.
+ *
+ * @constructor
+ */
+function History() {
+    this.sampleHistory = [];
+    this.iterator = 0;
+    this.length = this.sampleHistory.length;
+    this.lastSample = null;
+}
+
+/**
+ * add: Adds an item to the history.
+ *
+ * Any elements after the current iterator are removed.
+ *
+ * @this {History}
+ * @param {string} sampleId - The sampleID to add to the history.
+ */
+History.prototype.add = function(sampleId) {
+    if(sampleId !== this.lastSample) {
+        this.sampleHistory = this.sampleHistory.slice(0, this.iterator+1);
+        this.sampleHistory.push(sampleId);
+        this.iterator = this.sampleHistory.length - 1;
+        this.length = this.sampleHistory.length;
+        this.lastSample = sampleId;
```

Comment:
It's returning a subset of the array, overwriting the old reference with new array, then the garbage collector will see there's no more reference to the old array and remove it. You raise a good point though, it's not the most efficient way to do it. JS Arrays are actually hash tables with sequential integer keys (remember everything in JS inherits from the base Object class, so this makes sense in topsy turvy JS land..). So the slice operation is O(n), and the push is O(1).. doing it the way you suggested will mean overwriting is O(1), and the push is O(1).. so O(1).. yay! :)